### PR TITLE
Implement perturb no-op behaviour

### PR DESCRIPTION
**Pull request by cgarciae**
# What does this PR do?

* If `perturbations` are not passed, `perturb` behaves like a no-op. When the user is done inspecting the gradients it can omit passing `perturbations` without having to change the model.

### CODE REVIEW

File: flax/linen/module.py [hunk 1]
```diff
@@ -1655,8 +1655,18 @@ def loss(params, perturbations, inputs, targets):
...
+
+      model.apply({'params': params, 'perturbations': perturbations}, x) # works as expected
+      model.apply({'params': params}, x) # behaves like a no-op
+
     """
-    value += self.variable(collection, name, lambda: jnp.zeros_like(value)).value # type: ignore
+    def _root_has_collection():
+      assert self.scope is not None
+      return collection in self.scope.root._variables
+    if self.is_mutable_collection(collection) or _root_has_collection():
```
Comment (marcvanzee) [inline flax/linen/module.py:1668]: Maybe add a comment explaining this. I suppose the first clause handles initialization (since everything is mutable by default) and the second one apply (it will be ignored if you don't pass it).

Comment (cgarciae) [reply flax/linen/module.py:1672]: Done!

**Review by marcvanzee — APPROVED** [review]
Looks good!
